### PR TITLE
Added multi stage dockerfile to contain builds

### PR DESCRIPTION
**Pull request by mzedel**
While moving other parts of mender to contained docker builds, mender-artifact would also be a nice to have. Due to the lzma dependency this is not based on scratch or busybox, but alpine.

Changelog: None
Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

### CODE REVIEW

**Comment by kacf** [comment]
The Travis error is being fixed in #189.

File: Makefile [hunk 1]
```diff
@@ -24,7 +24,7 @@ TOOLS = \
 VERSION = $(shell git describe --tags --dirty --exact-match 2>/dev/null || git rev-parse --short HEAD)
 
 GO_LDFLAGS = \
-	-ldflags "-extldflags '-static' -X main.Version=$(VERSION)"
+	-ldflags "-X main.Version=$(VERSION)"
```
Comment (kacf) [inline Makefile]: Out of curiosity: Why did you remove this?

I saw the warnings when this was readded, but it worked for me, and this could be quite desirable, especially on Mac and Windows, where the the lzma library is less likely to exist.

Comment (mzedel) [reply Makefile]: here I just removed it to make as minimal a change as possible to the way it works now, as I didn't want to cause a breaking change somewhere else. It worked in the container as well but I wasn't sure about the warnings...

**Comment by mzedel** [comment]
@kacf to be able to extract the binary afterwards I had to use a base image with a shell (just the `sh`) (another reason not to build this from `scratch`) - then I can run the image with the shell instead and mount a volume where the binary should go, and copy the binary out of the container. And since the shell has nothing else to execute, the shell and with it the container shuts down as soon as the copy finishes.
This way the container should be able to be run almost like an executable and we have a unified container for build artifacts...

File: Makefile [hunk 2]
```diff
@@ -42,6 +42,16 @@ endif
 build:
 	$(GO) build $(GO_LDFLAGS) $(BUILDV) $(BUILDTAGS) -o $(PKGNAME) $(BUILDFILES)
 
+build-contained:
+	if [ -f mender-artifact ]; \
+	then \
+		echo "mender-artifact binary already exists, please move or rename"; \
```
Comment (lluiscampos) [inline Makefile]: Why don't we just remove unconditionally mender-artifact binary?

File: Makefile [hunk 3]
```diff
@@ -42,6 +42,16 @@ endif
...
+		exit 1; \
+	fi; \
+	image_id=$$(docker build -f Dockerfile . | awk '/Successfully built/{print $$NF;}'); \
```
Comment (lluiscampos) [inline Makefile]: For getting image_id, isn't in easier to `--tag` the build to a known name?

File: Makefile [hunk 4]
```diff
@@ -42,6 +42,16 @@ endif
 build:
 	$(GO) build $(GO_LDFLAGS) $(BUILDV) $(BUILDTAGS) -o $(PKGNAME) $(BUILDFILES)
 
+build-contained:
```
Comment (lluiscampos) [inline Makefile]: It is a nice addition to our Makefile ;)

File: Makefile (hunk 3 above)
Comment (mzedel) [reply Makefile]: that might overwrite an existing tag and I wanted to avoid that... - will still remove an existing image with the same hash

**Comment by mzedel** [comment]
@kacf now I finally got around the make differences and added the contained build...

File: Makefile (hunk 2 above)
Comment (mzedel) [reply Makefile]: good question... would be in line with the regular `build` task...

**Comment by lluiscampos** [comment]
The test failure can be ignored, it is this spurious server error that we see some times.

As @kacf is on vacation, should I approve this?

Replace the if/else for an unconditional remove of the mender-artifact in the `build-contained` target of the `Makefile` and I'll be happy to do so.

File: Makefile [hunk 5]
```diff
@@ -42,6 +42,15 @@ endif
...
+	if [ -f mender-artifact ]; \
+	then \
+		rm mender-artifact; \
+	fi; \
```
Comment (lluiscampos) [inline Makefile]: You can simplify this even more by just running `rm -f mender-artifact`.

IMO, the least conditionals and for loops in Makefiles, the better. Only the strictly needed ones.

**Comment by lluiscampos** [comment]
I dropped a new comment, sorry. Other than that looks good and I am approving it.

I don't think it is important to wait for the tests, as we had a pass last week. So feel free to merge it if it is blocking you.

File: Makefile (hunk 5 above)
Comment (mzedel) [reply Makefile]: my bad, I just thought of either regular `rm` (which potentially results in a warning) or reusing the condition... is `rm -f` now

**Comment by lluiscampos** [comment]
Last commit is not signed.

Also, in the client team we try to group the changes in as little commits as possible. Rule of thumb is to rebase commits that amend previous commits (i.e. squash them), and in general keep the change set consistent with its log message.

Can you please tidy up a bit?

Thanks

**Comment by mzedel** [comment]
@lluiscampos now I squashed the last commits and fixed the signoff. Shall I squash some more?

**Comment by lluiscampos** [comment]
It is good :+1: 

Ignore the build errors and merge.
